### PR TITLE
Fixed an issue with parallel transition sources not being reentered correctly

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/statelyai/xstate/issues/4360

### CODE REVIEW

File: packages/core/src/stateUtils.ts [hunk 1]
```diff
@@ -1381,15 +1388,16 @@ function addDescendantStatesToEnter<
...
+  ancestors: AnyStateNode[],
+  reentrancyDomain?: AnyStateNode
 ) {
-  const properAncestors = getProperAncestors(stateNode, toStateNode);
-  for (const anc of properAncestors) {
-    statesToEnter.add(anc);
+  for (const anc of ancestors) {
+    if (!reentrancyDomain || isDescendant(anc, reentrancyDomain)) {
+      statesToEnter.add(anc);
+    }
```
Comment (Andarist) [inline packages/core/src/stateUtils.ts:1400]: I'm not a fan of adding this particular logic here (and the related code in the caller) but I wasn't able to get rid of this so far.

File: packages/core/test/parallel.test.ts [hunk 2]
```diff
@@ -1172,4 +1172,107 @@ describe('parallel states', () => {
 
     expect(service.getSnapshot().value).toBe('finished');
   });
+
+  it('source parallel region should be reentered when a transition within it targets another parallel region (parallel root)', async () => {
+    const machine = createMachine({
```
Comment (davidkpiano) [inline packages/core/test/parallel.test.ts:1177]: CleanShot 2023-11-02 at 08 02 37@2x

File: packages/core/test/parallel.test.ts [hunk 3]
```diff
@@ -1172,4 +1172,107 @@ describe('parallel states', () => {
...
+      'exit: Operation',
+      'enter: Operation',
+      'enter: Operation.Waiting',
+      'enter: Mode',
+      'enter: Mode.Demo'
+    ]);
+  });
+
+  it('source parallel region should be reentered when a transition within it targets another parallel region (nested parallel)', async () => {
+    const machine = createMachine({
```
Comment (davidkpiano) [inline packages/core/test/parallel.test.ts:1226]: CleanShot 2023-11-02 at 08 03 06@2x

File: packages/core/test/actions.test.ts [hunk 4]
```diff
@@ -2021,6 +2021,41 @@ describe('initial actions', () => {
       ]
     `);
   });
+
+  it('should execute actions of the initial transition when taking a root reentering self-transition', () => {
+    const spy = jest.fn();
+    const machine = createMachine({
```
Comment (davidkpiano) [inline packages/core/test/actions.test.ts:2027]: CleanShot 2023-11-02 at 08 06 39@2x

**Review by davidkpiano — APPROVED** [review]

**Pull request merged**